cpu/sam0_common/periph/dma.c

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (C) 2020 Koen Zandberg <koen@bergzand.net>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_sam0_common
+ * @{
+ *
+ * @file
+ * @brief       Low-level DMA driver implementation
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ *
+ * @}
+ */
+
+#include "periph_cpu.h"
+#include "periph_conf.h"
+#include "mutex.h"
+#include "assert.h"
+#include "bitarithm.h"
+#include "pm_layered.h"
+#include "thread_flags.h"
+#include "periph/gpio.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#ifndef CONFIG_DMA_NUMOF
+#define CONFIG_DMA_NUMOF DMAC_CH_NUM
+#endif
+
+/* In memory DMA transfer descriptors */
+static DmacDescriptor descriptors[CONFIG_DMA_NUMOF];
+static DmacDescriptor writeback[CONFIG_DMA_NUMOF];
+
+/* Bitmap of dma channels available */
+static uint32_t channels_free = (1LLU << CONFIG_DMA_NUMOF) - 1;
+
+struct dma_ctx {
+    mutex_t sync_lock;
+};
+
+struct dma_ctx dma_ctx[CONFIG_DMA_NUMOF];
+
+static void _poweron(void)
+{
+#if defined(MCLK)
+    MCLK->AHBMASK.reg |= MCLK_AHBMASK_DMAC;
+#else
+    PM->AHBMASK.reg |= PM_AHBMASK_DMAC;
+    PM->APBBMASK.reg |= PM_APBBMASK_DMAC;
+#endif
+}
+
+void dma_init(void)
+{
+    _poweron();
+    for (unsigned i = 0; i < CONFIG_DMA_NUMOF; i++) {
+        mutex_init(&dma_ctx[i].sync_lock);
+    }
+    /* Enable all priorities with RR scheduling */
+    DMAC->CTRL.reg = DMAC_CTRL_LVLEN0 |
+                     DMAC_CTRL_LVLEN1 |
+                     DMAC_CTRL_LVLEN2 |
+                     DMAC_CTRL_LVLEN3;
```

Comment:
Just a matter of personal taste, but I generally find 

```suggestion
    DMAC->CTRL.reg = DMAC_CTRL_LVLEN0
                   | DMAC_CTRL_LVLEN1
                   | DMAC_CTRL_LVLEN2
                   | DMAC_CTRL_LVLEN3;
```

to be more visually pleasing.
Feel free to ignore if you disagree - in that case, just squash directly.